src/Microsoft.VisualStudio.Threading.Tests/TestBase.cs

```diff
@@ -290,53 +291,37 @@ protected void ExecuteOnDispatcher(Action action)
             });
         }
 
-        protected void ExecuteOnDispatcher(Func<Task> action, bool staRequired = true)
+        protected void ExecuteOnDispatcher(Func<Task> action)
         {
-            Action worker = delegate
+            if (!SingleThreadedTestSynchronizationContext.IsSingleThreadedSyncContext(SynchronizationContext.Current))
             {
-                var frame = SingleThreadedTestSynchronizationContext.NewFrame();
-                Exception failure = null;
-                SynchronizationContext.Current.Post(
-                    async _ =>
-                    {
-                        try
-                        {
-                            await action();
-                        }
-                        catch (Exception ex)
-                        {
-                            failure = ex;
-                        }
-                        finally
-                        {
-                            frame.Continue = false;
-                        }
-                    },
-                    null);
-
-                SingleThreadedTestSynchronizationContext.PushFrame(SynchronizationContext.Current, frame);
-                if (failure != null)
-                {
-                    ExceptionDispatchInfo.Capture(failure).Throw();
-                }
-            };
-
-            if ((!staRequired || Thread.CurrentThread.GetApartmentState() == ApartmentState.STA) &&
-                SingleThreadedTestSynchronizationContext.IsSingleThreadedSyncContext(SynchronizationContext.Current))
-            {
-                worker();
+                SynchronizationContext.SetSynchronizationContext(SingleThreadedTestSynchronizationContext.New());
```

Comment:
do we need restore the original context before leaving this function?

Reply:
This method was already doing this (in almost every case). So this is parity behavior. 
This might be an area that needs cleanup. Honestly though we need to clean up some more of these tests is a significant way and touching little utility functions like this has turned out to be hazardous to test stability, so I'm not eager to make further changes as part of this PR.